TestStage1LoadingFromFlags fails on Ubuntu 16.04

The test `TestStage1LoadingFromFlags` passes on Fedora-22, CentOS-7, Debian-8.1 and Ubuntu-15.10 but fails on Ubuntu 16.04:

It seems to be reproducible if I run all the `TestStage1Loading*` tests at the same time but not if I run only `TestStage1LoadingFromFlags`:
```
make functional-check GO_TEST_FUNC_ARGS='-run  TestStage1Loading'
```

```
=== RUN   TestStage1LoadingFromFlags
--- FAIL: TestStage1LoadingFromFlags (7.91s)
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-421866435 --local-config=/tmp/localdir-937596998 --system-config=/tmp/systemdir-815087853 --insecure-options=image,tls --debug run --stage1-path="../../../../../../tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-421866435 --local-config=/tmp/localdir-937596998 --system-config=/tmp/systemdir-815087853 --insecure-options=image,tls --debug run --stage1-path="../../../../../../tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-572286824 --local-config=/tmp/localdir-482576295 --system-config=/tmp/systemdir-566306522 --insecure-options=image,tls --debug run --stage1-path="/var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-572286824 --local-config=/tmp/localdir-482576295 --system-config=/tmp/systemdir-566306522 --insecure-options=image,tls --debug run --stage1-path="/var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_stage1_loading_test.go:368: skipping the testcase with the flag --stage1-url=http://..., reason - tested in TestStage1LoadingFromFlagsHttp
	rkt_stage1_loading_test.go:368: skipping the testcase with the flag --stage1-url=docker://..., reason - tested in TestStage1LoadingFromFlagsDocker
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-020081265 --local-config=/tmp/localdir-259598108 --system-config=/tmp/systemdir-484906955 --insecure-options=image,tls --debug run --stage1-url="https://[::1]:443/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:583: server: Trying to serve "/rkt-stub-stage1.aci"
	rkt_tests.go:583: server:   done.
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-020081265 --local-config=/tmp/localdir-259598108 --system-config=/tmp/systemdir-484906955 --insecure-options=image,tls --debug run --stage1-url="https://[::1]:443/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-444140206 --local-config=/tmp/localdir-654537781 --system-config=/tmp/systemdir-958316048 --insecure-options=image,tls --debug run --stage1-url="file:///var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-444140206 --local-config=/tmp/localdir-654537781 --system-config=/tmp/systemdir-958316048 --insecure-options=image,tls --debug run --stage1-url="file:///var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-stub-stage1.aci" /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_tests.go:113: Running command: /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt --dir=/tmp/datadir-064868911 --local-config=/tmp/localdir-208232898 --system-config=/tmp/systemdir-722693177 --insecure-options=image,tls --debug run --stage1-name=localhost/rkt-stub-stage1:0.0.1 /var/tmp/rkt/builds/build-rkt-coreos-/build-rkt-0.16.0+git/tmp/functional/rkt-inspect.aci
	rkt_stage1_loading_test.go:602: Did not get the expected string "success, stub stage1 would at this point switch to stage2":
		image: searching for app image localhost/rkt-stub-stage1
		run: 
		  └─discovery failed for "localhost/rkt-stub-stage1"
		    └─Get https://localhost?ac-discovery=1: dial tcp 127.0.0.1:443: getsockopt: connection refused
	rkt_tests.go:127: rkt didn't terminate correctly: exit status 1
		Output:
```

--- a/tests/testutils/aci-server/server.go
+++ b/tests/testutils/aci-server/server.go
@@ -25,6 +25,7 @@ import (
 	"net/http/httptest"
 	"path/filepath"
 	"strings"
+	"time"
 )
 
 type PortType int
@@ -369,7 +370,6 @@ func NewServer(setup *ServerSetup) *Server {
 	return server
 }
 
-// based on code from net/http/httptest
 func newLocalListener(port PortType, protocol ProtocolType) net.Listener {
 	portNumber := 0
 	if port == PortFixed {
@@ -380,13 +380,31 @@ func newLocalListener(port PortType, protocol ProtocolType) net.Listener {
 			portNumber = 443
 		}
 	}
-	l, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", portNumber))
+	addrs, err := net.LookupHost("localhost")
 	if err != nil {
-		if l, err = net.Listen("tcp6", fmt.Sprintf("[::1]:%d", portNumber)); err != nil {
-			panic(fmt.Sprintf("aci test server: failed to listen on a port: %v", err))
-		}
+		panic(`aci test server: failed to look up "localhost", really`)
 	}
-	return l
+	var lookupErrs []string
+	for try := 0; try < 2; try++ {
+		for _, addr := range addrs {
+			addrport := fmt.Sprintf("%s:%d", addr, portNumber)
+			l, err := net.Listen("tcp", addrport)
+			if err == nil {
+				return l
+			}
+			lookupErrs = append(lookupErrs, fmt.Sprintf("(listen on %s, attempt #%d: %v)", addrport, try+1, err))
+		}
+		// TODO: When we have discovery on a custom port then
+		// we could drop listening on fixed ports, so we
+		// probably won't get any races between old server
+		// stopping to listen and new server starting to
+		// listen.
+		// https://github.com/appc/spec/pull/110
+		// Might be possible with ABD:
+		// https://github.com/appc/abd
+		time.Sleep(time.Second)
+	}
+	panic(fmt.Sprintf("aci test server: failed to listen on localhost:%d: %v", portNumber, lookupErrs))
 }
 
 func sprintCreds(host, auth, creds string) string {

